### PR TITLE
Immunotherapy Tensor Maps and Bootstrapping Notebooks

### LINKED ISSUE #1025: Make tensor maps for cardiovascular events after immunotherapy
## What and why
So we can model cardiovascular events after ICI from the pre-treatment ECG.

## Solution(s)
Create tensor maps in `/ml4c3/tensormap/immunotherapy.py` that parse data in `~/dropbox/ecg-immunotherapy/data/filename.csv`.

## Acceptance criteria
```
--input_tensors ecg_2500_zscore_no_pacemaker_180_days_pre_ici
--output_tensors ici_cv_event_180_days_post_ecg
```

`tensormaps/immunotherapy.py`

## Blocked by or pending

### LINKED ISSUE #1080: Create stratified splits for ECGNet-ICI

### LINKED ISSUE #1084: Create c3po pre-training cohort that excludes STS, ICI, and echo MRNs

### PR DESCRIPTION
**Pull request by StevenSong**
## Issues closed
Closes #1025 
Closes #1080 
Closes #1084 
Enables #1026 

## Changelog
- [x] cohort name is `ici` (in dropbox, data file is symlinked to ici.csv)
- [x] new ici tmaps/definitions for continuous/binary/categorical features
- [x] given `ici_cv_event`, `ici_cv_event_case` will restrict to only case patients 
- [x] update updaters
- [x] adds bootstrapping notebook for ici
- [x] adds notebook to create training splits for c3po, excluding echo, sts, ici

### CODE REVIEW

File: tensormap/ici.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        tensor = np.zeros((len(values), len(tm.channel_map)))
+        for i, value in enumerate(values):
+            tensor[i, tm.channel_map[value]] = 1
+        return tensor
+
+    return tensor_from_file
+
+
+def _make_regex(keys: Iterable[str]) -> str:
+    return fr"^({'|'.join(keys)})(_case)?"
```
Comment (erikr) [inline tensormap/ici.py:171]: This is a cool approach.

**Review by erikr — APPROVED** [review]
Great work. Minor questions.

Just one quip -- everything feels so complicated now. I think algorithmically creating tensor maps is the right approach given how many there are, and the flexibility it affords with setting parameters from the TMap name. But I wish we could make it simpler.

File: tensormap/ici.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    return tensor_from_file
+
+
+def make_binary_tensor_from_file(
+    key: str,
+) -> Callable[[TensorMap, PatientData], np.ndarray]:
+    def tensor_from_file(tm: TensorMap, data: PatientData) -> np.ndarray:
+        ici_dates = tm.time_series_filter(data)
+        positive_values = data[ICI_PREFIX].loc[ici_dates.index, key].to_numpy()
+        negative_values = 1 - positive_values
```
[the remark this answers is not in the document]
Comment (StevenSong) [reply tensormap/ici.py:149]: `positive_values` is like `cv_event` which may be `0` or `1`

then the opposite `negative_values` is like `no_cv_event` which ought to be `1` or `0`, respectively

1 - 0 = 1
1 - 1 = 0

basically just to create the 1 hot encoding for binary label

File: tensormap/ici.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    ),
+    "ici_type": (
+        "ici_type",
+        {
+            "none": 0,
+            "PD1": 1,
+            "PDL1": 2,
+            "CTLA4": 3,
+            "CTLA4 AND PD1": 4,
+            "CTLA4 OR PD1": 5,
```
[the remark this answers is not in the document]
Comment (StevenSong) [reply tensormap/ici.py:114]: lets make this a discussion/question

[the remark this answers is not in the document]
Comment (StevenSong) [reply tensormap/ici.py:114]: https://github.com/aguirre-lab/ml4c3/discussions/1112

**Pull request merged**